src/analysis_and_optimization/Optimize.ml

```diff
@@ -114,30 +134,41 @@ let subst_args_stmt args es =
    then to check if that flag is set after each loop. Then, if a 'return' break
    is called from an inner loop, there's a cascade of breaks all the way out of
    the dummy loop. *)
-let handle_early_returns opt_var b =
-  let returned = Gensym.generate ~prefix:"inline_" () in
-  let f = function
+let handle_early_returns (fname : string) opt_var stmt =
+  let returned =
+    Gensym.generate ~prefix:("inline_" ^ fname ^ "_early_ret_check_") () in
+  let generate_inner_breaks num_returns stmt_pattern =
+    match stmt_pattern with
     | Stmt.Fixed.Pattern.Return opt_ret -> (
       match (opt_var, opt_ret) with
       | None, None -> Stmt.Fixed.Pattern.Break
       | Some name, Some e ->
-          SList
-            [ Stmt.Fixed.
-                { pattern=
-                    Assignment
-                      ( (returned, UInt, [])
-                      , Expr.Fixed.
-                          { pattern= Lit (Int, "1")
-                          ; meta=
-                              Expr.Typed.Meta.
-                                { type_= UInt
-                                ; adlevel= DataOnly
-                                ; loc= Location_span.empty } } )
-                ; meta= Location_span.empty }
-            ; Stmt.Fixed.
-                { pattern= Assignment ((name, Expr.Typed.type_of e, []), e)
-                ; meta= Location_span.empty }
-            ; {pattern= Break; meta= Location_span.empty} ]
+          if num_returns > 1 then
```

Comment:
I'd prefer this be a guard clause in the match rather than this big if, so you have two branches:
```ocaml
| Some name, Some e when num_returns > 1 -> ... contents of if
| Some name, Some e -> ... contents of else
```

Similar for the other cases where the body of the match is just an if statement

Reply:
Much cleaner!